bigframes/session/__init__.py

```diff
@@ -758,9 +759,31 @@ def _read_gbq_table(
 
         # Use a time travel to make sure the DataFrame is deterministic, even
         # if the underlying table changes.
-        # TODO(b/340540991): If a dry run query fails with time travel but
+
+        # If a dry run query fails with time travel but
         # succeeds without it, omit the time travel clause and raise a warning
         # about potential non-determinism if the underlying tables are modified.
+        sql = bigframes.session._io.bigquery.to_query(
+            f"{table_ref.project}.{table_ref.dataset_id}.{table_ref.table_id}",
+            index_cols=index_cols,
+            columns=columns,
+            filters=filters,
+            time_travel_timestamp=time_travel_timestamp,
+            max_results=None,
+        )
+        dry_run_config = bigquery.QueryJobConfig()
+        dry_run_config.dry_run = True
+        try:
+            self._start_query(sql, job_config=dry_run_config)
+        except google.api_core.exceptions.NotFound:
+            time_travel_timestamp = None
+            warnings.warn(
```

Comment:
Let's use a custom category here from `bigframes.exceptions` so folks can filter this out in pipelines where they know they are fine with a query on a table that could be changing from under them.